### PR TITLE
Generate index mth

### PR DESCRIPTION
**Pull request by cyisfor**
Minetest's "remote media" stuff is a neat idea, and it lets you name the files after their sha1 hash, rather than use "content_dirt.png" or whatnot. That way you can update what dirt looks like, without clients keeping the old texture cached. 

But you cannot do that, without an index.mth file in the media directory. There doesn't seem any way to generate index.mth. Like... at all. Anywhere. 

I'm a terrible hack, but I thought there should at least be _some_ way to do it. I noticed the client generates a file of the _format_ index.mth is in, but only sends "files we don't have" to the server, rather than keeping a list of files we do have. So I wrote a quick little C++ program that uses the serialization algorithm in clientmedia.cpp to generate an index.mth from a list of hex encoded hashes on stdin. 

The idea is you'd clear your cache, connect to your own server without remote media enabled, then 

``` sh
ls cache/media | ./bin/generate_index_mth > cache/media/index.mth
```

Then you can symlink that directory from your web server's docroot, and provide the URL to that as a remote_media URL.

A better solution might be for clients to just store and update an index.mth when they download a bunch of stuff to the cache. I didn't want to take a list of "content_dirt.png" files as input, or walk the tree looking for textures/ sounds/ stuff/ etc, because hashing hundreds of files is expensive, and it's easier to just copy cache/media/ which already stores them by their hashed name. Ideally, you'd track which "content_dirt.png" filename corresponds to which hash, and when that file is newer than the hashed file, you delete the hashed file, and re-hash the named file.

...but `ls cache/media | ./bin/generate_index_mth > cache/media/index.mth` is easier.

### CODE REVIEW

**Comment by cyisfor** [comment]
Oh, I tested it and it works perfectly for me, with a bunch of mods installed.

File: src/clientmedia.cpp [hunk 1]
```diff
@@ -483,7 +485,10 @@ void ClientMediaDownloader::startRemoteMediaTransfers()
 
 void ClientMediaDownloader::startConventionalTransfers(Client *client)
 {
-	assert(m_httpfetch_active == 0);	// pre-condition
+#ifdef SERVER
+  throw "fail";
```
Comment (SmallJoker) [inline src/clientmedia.cpp]: Use tabs to indent. More important: Why is this part required?

File: src/generate_index.mth.cpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+#include <iostream>
+#include <string>
+using namespace std;
+
+int main(int argc, char *argv[])
+{
+  string hhash;
+  ClientMediaDownloader store;
+  while(!cin.eof()) {
```
Comment (SmallJoker) [inline src/generate_index.mth.cpp]: Various style errors, please read http://dev.minetest.net/Code_style_guidelines

**Comment by rubenwardy** [comment]
This script is currently used: https://gist.github.com/sfan5/6351560

**Comment by PilzAdam** [comment]
@rubenwardy this should be documented somewhere.

**Comment by rubenwardy** [comment]
Definitely. It also seems a little broken recently.

**Comment by sfan5** [comment]
It would be good to have this utility but your code seems is a terrible hack (as you already said) and won't be merged like this.

What's correct is that the way of generating index.mth should be documented somewhere

**Comment by ShadowNinja** [comment]
I made this utility: https://github.com/ShadowNinja/mt_media_collector/
Example usage:
```
cargo build --release
target/release/mt_media_collecter --copy \
        --out /srv/http/mt_media \
        --world ~/.minetest/worlds/foo \
        --game /usr/share/minetest/games/minetest_game \
        ~/.minetest/mods /other/mod/dirs
```

**Comment by sofar** [comment]
Full implementation that removes the need to anything:

github.com/sofar/mtmediasrv

- collects all media from mods, games
- creates proper webroot content
- dynamically handles client POST data and returns correct hash info
- does not need a static index.mth
- functions as a FCGI service
- Linux binary builds available - can be run on almost any linux distro out of the box
